### ISSUE TITLE
PCDATA invalid Char value

### ISSUE THREAD

**Issue by Carmezim** [MasterScrat/Chatistics#4]
Had this issue when running the parser and to solve is just add proper encoding on line 36 if anyone happens to face the same problem:
```
archive = etree.parse(filePath + "/" + filename,
                            etree.XMLParser(encoding='utf-8',
                                            ns_clean=True,
                                            recover=True))
```

should submit a PR?

**Comment by MasterScrat** [comment]
Setting the encoding is good but I don't like the `recover` flag (= "try hard to parse through broken XML"): I'd rather have the parser crash than misinterpret data.

Does it work for you without the `recover` flag? Also could you share an example of logs that cause the problem in the first place?

**Comment by Carmezim** [comment]
Sure, stack trace from the original code:
```
Traceback (most recent call last):
  File "parse_test.py", line 36, in <module>
    archive = etree.parse(filePath + "/" + filename)
  File "src/lxml/lxml.etree.pyx", line 3427, in lxml.etree.parse (src/lxml/lxml.etree.c:85131)
  File "src/lxml/parser.pxi", line 1782, in lxml.etree._parseDocument (src/lxml/lxml.etree.c:124005)
  File "src/lxml/parser.pxi", line 1808, in lxml.etree._parseDocumentFromURL (src/lxml/lxml.etree.c:124374)
  File "src/lxml/parser.pxi", line 1712, in lxml.etree._parseDocFromFile (src/lxml/lxml.etree.c:123169)
  File "src/lxml/parser.pxi", line 1115, in lxml.etree._BaseParser._parseDocFromFile (src/lxml/lxml.etree.c:117533)
  File "src/lxml/parser.pxi", line 573, in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:110510)
  File "src/lxml/parser.pxi", line 683, in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:112276)
  File "src/lxml/parser.pxi", line 613, in lxml.etree._raiseParseError (src/lxml/lxml.etree.c:111124)
lxml.etree.XMLSyntaxError: PCDATA invalid Char value 28, line 117, column 5169
```

With `recover=False`:
```
Traceback (most recent call last):
  File "parse_messenger.py", line 36, in <module>
    archive = etree.parse(filePath + "/" + filename, etree.XMLParser(encoding='utf-8', ns_clean=True, recover=False))
  File "src/lxml/lxml.etree.pyx", line 3427, in lxml.etree.parse (src/lxml/lxml.etree.c:85131)
  File "src/lxml/parser.pxi", line 1782, in lxml.etree._parseDocument (src/lxml/lxml.etree.c:124005)
  File "src/lxml/parser.pxi", line 1808, in lxml.etree._parseDocumentFromURL (src/lxml/lxml.etree.c:124374)
  File "src/lxml/parser.pxi", line 1712, in lxml.etree._parseDocFromFile (src/lxml/lxml.etree.c:123169)
  File "src/lxml/parser.pxi", line 1115, in lxml.etree._BaseParser._parseDocFromFile (src/lxml/lxml.etree.c:117533)
  File "src/lxml/parser.pxi", line 573, in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:110510)
  File "src/lxml/parser.pxi", line 683, in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:112276)
  File "src/lxml/parser.pxi", line 613, in lxml.etree._raiseParseError (src/lxml/lxml.etree.c:111124)
lxml.etree.XMLSyntaxError: PCDATA invalid Char value 28, line 117, column 5169
```

**Comment by MasterScrat** [comment]
Ah so it looks like the file is actually broken :-/

Can you find in the HTML file what the problem is on line 117, column 5169? maybe some special character?

**Comment by Carmezim** [comment]
Oh of course it is haha my bad, sorry about that. No worries, I'll take a look but this can be closed.

Thanks!

**Comment by MasterScrat** [comment]
Well if the file you got from Facebook is broken then probably other people will have the same problem as well, so we need a way to handle this situation...

**Comment by Carmezim** [comment]
It's a DC4 (device control four) character (`ord(\x14)`)

**Comment by Carmezim** [comment]
It's an ASCII control character, file separator (ord('\x1c') -> Char value 28) invalid in xml

**Comment by Carmezim** [comment]
Yes, it's a message. By the stack trace, the file 286 line 117 despite the column number not making sense I think it might a tilde on top of an "o": õ. It's the only weird character on line 117.

**Comment by Carmezim** [comment]
Yes it's a message.

edit: the file suddenly breaks into this weird formatting so it's not the "õ" as also it doesn't correspond with the char it got mad about. So the file separator might make sense? I left some of the surroundings to give a notion
```
<span class="user">Facebook User</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div><p><redacted></p><div class="message"><div class="message_header"><span class="user">Facebook User</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div><p><redacted></p><div class="message"><div class="message_header"><span class="user">Adriano Carmezim</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div><p>OH
<span class="user">Adriano Carmezim</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div><p>OH
3 sugestões:
1 - <redacted>
2 - <redacted>
3 - Bob </p>
</p><div class="message"><div class="message_header"><span class="user">Facebook User</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div><p><redacted></p><div class="message"><div class="message_header"><span class="user">Facebook User</span><span class="meta">Thursday, February 14, 2013 at 11:25pm UTC</span></div></div>
```

**Comment by MasterScrat** [comment]
Ok, I guess we have to assume Facebook's files may contain weird characters then.

I don't see a better option than using `recover`,  I added it in ed4ac96.

Thanks for the report!

**Comment by JccSanabria** [comment]
I had the same issue converting pulled data from ldap in json to xml. Thanks for that recover!
